### ISSUE TITLE
Permanent link is not fully permanent

### ISSUE THREAD

**Issue by martinpopel** [ufal/perl-pmltq-web#162]
https://lindat.mff.cuni.cz/services/pmltq has a "Permanent link" button which suggests the result of a given query will be always the same. Unfortunately, the results are sometimes returned in a different order. If random shuffling of the order is a feature, can we turn it off for permanent links?

**Comment by martinpopel** [comment]
I understand that permanent links specify just the treebank and query string. But when users see "permanent" they have some expectations. It is even trickier because sometimes the query returns the results in the same order (when trying to run the same query in the same browser with the same limit within a short time after the previous execution), so the users may think that also the result is permanent. For example, I have spent some time describing [PDT-C errors](https://wiki.ufal.ms.mff.cuni.cz/internal:pcdtc:chyby) such as "the sixth sentence found by this query is...".

So there are two possible solutions:
* Warn the users that the results are not permanent whenever they use the permanent link button.
* Make also the results permanent, which means just fixing the non-deterministic order because the treebanks are versioned and the PML-TQ language does not evolve much.

> Queries without output filters are not possible to sort. 

I am no PostgreSQL expert, so maybe there is an easier way to prevent the non-determinism, but what about adding `ORDER BY` to each PostgreSQL query? And using e.g. IDs of all nodes mentioned in the query? This should make the order deterministic.

If there is currently no guarantee about the ordering (and no way to specify the order within the query), no users should be disappointed by fixing the ordering.

Some users may appreciate random shuffling of the query results, but it should be optional and replicable (i.e. stable sort), such as in [KonText](https://wiki.korpus.cz/doku.php/manualy:kontext:konkordance#promichat).

**Comment by martinpopel** [comment]
OK, so adding ORDER BY by default to all queries is not a good idea. But unless there is another way to prevent the non-determinism (which seems to be caused by PostgreSQL rather than Perl if a different set of 100 results is returned each time with LIMIT 100), it would be nice to at least have an option to turn on the ordering. When I know there are less than 100 
(or 1000) results, the ordering would cause no significant slowdown (if I want all of them in the permanent query result anyway).

**Comment by stranak** [comment]
So maybe this is not so much about the "permanent link" button, but rather making it clear in the docs that the order of results is not guaranteed due to the database / technology in general? 

The most I see we could do is make that information easy to reach.

**Comment by martinpopel** [comment]
Yes, it is not *so much* about the permanent link button - it would be nice to have deterministic order for the same query even if not using the button. But the as I wrote the users may think that also the result is permanent when they see "permanent link". So if we cannot fix the non-determinism bug, we should warn the users whenever they use the permanent link button.

**Comment by stranak** [comment]
“Permanent / persistent query link”?

**Comment by martinpopel** [comment]
Changing the name is not enough. My suggestion is to add the following notice (or similar) somewhere near the "Press ctrl + c to copy":
"The permanent link encodes just the query and treebank. Ordering of the returned results may change."

**Comment by stranak** [comment]
You are still mixing 2 things and I am not convinced it is a good idea. The change in working makes it clear that the link is to the query, not the results. Another issue is how pmltq works. A side-effect of it is the impermanent ordering of results. I am not convinced we need to make a strong warning about that in every place that has to do with queries. After all, it seems we get this question about once in a decade. 

Thus I propose to do 2 things: 
  - make sure that it is clear the link is to the query,
  - make clear in the documentation how and why the results are returned regarding order

Pavel
